### PR TITLE
feat(lsp): jump to diagnostics by position

### LINKED ISSUE #13915: vim.lsp.diagnostic.goto_next/goto_prev jump to next/previous line even if the current line has additional diagnostics
- `nvim --version`:
```
NVIM v0.5.0-dev+1080-g971e0ca90
Build type: Release
LuaJIT 2.0.5
Compilation: /usr/bin/cc -D_FORTIFY_SOURCE=2 -march=native -O2 -pipe -fno-plt -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -O2 -DNDEBUG -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wmissing-prototypes -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fno-common -fdiagnostics-color=always -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -DMIN_LOG_LEVEL=3 -I/tmp/yaytmp-1000/neovim-git/src/build/config -I/tmp/yaytmp-1000/neovim-git/src/neovim-git/src -I/usr/include -I/tmp/yaytmp-1000/neovim-git/src/build/src/nvim/auto -I/tmp/yaytmp-1000/neovim-git/src/build/include
Compiled by wil@tesla

Features: +acl +iconv +tui
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/share/nvim"

Run :checkhealth for more info
```

### Steps to reproduce
1. Install `pyright` (https://github.com/microsoft/pyright)
2. Setup LSP + `pyright` using settings from https://github.com/neovim/nvim-lspconfig
3. Open a new Python file, create multiple errors on a single line, as well as errors on a later line
4. Run `:lua vim.lsp.diagnostic.goto_next()` and observe that the cursor jumps between lines, but not between diagnostics on the same line.

### Actual behaviour
The cursor jumps between diagnostics on distinct lines

### Expected behaviour
The cursor jumps between all diagnostics in sequence, including those on the same line as other diagnostics

### Preliminary investigation notes

I haven't had time to figure out the right way to fix this yet, but this seems to happen because `get_next`/`get_prev` (in `runtime/lua/vim/lsp/diagnostic.lua`) only pass the current line to `_iter_diagnostic_lines`, ignoring the column of the cursor position. Fixing this is slightly complicated because it appears that `get_prev` relies on decrementing this line value by 2 to search "backward"; naively continuing this will not have the desired behavior (i.e. this will continue to skip previous diagnostics on the same line as the current diagnostic).

### PR DESCRIPTION
**Pull request by pohzipohzi**
tries to fix #13915

before:

line

after:

linecol

### CODE REVIEW

File: runtime/lua/vim/lsp/diagnostic.lua
Comment (mjlbach) [inline runtime/lua/vim/lsp/diagnostic.lua:520]: I don't think it makes sense anymore for this function to be iter_diagnostic_lines, I think we should probably rename it something like iter_diagnostic_position, but it might make sense to have an iter_diagnostic_lines function as well (instead of replacing the old one).

File: runtime/lua/vim/lsp/diagnostic.lua
Comment (mjlbach) [inline runtime/lua/vim/lsp/diagnostic.lua:389]: I would refactor this to take a position instead of "cursor" and rename the M.get_position_diagnostics, then default as you have currently to using the current win position.

File: runtime/lua/vim/lsp/diagnostic.lua
Comment (mjlbach) [reply runtime/lua/vim/lsp/diagnostic.lua:520]: Maybe the new function should be iter_diagnostic_range, and instead of having a "search_forward" param have a start/end param that correspond to the buf position.

File: runtime/lua/vim/lsp/diagnostic.lua
Comment (mjlbach) [inline runtime/lua/vim/lsp/diagnostic.lua:1195]: same comment as above, I would make this "show_position_diagnostics"

File: runtime/lua/vim/lsp/diagnostic.lua
Comment (pohzipohzi) [reply runtime/lua/vim/lsp/diagnostic.lua]: > I don't think it makes sense anymore for this function to be iter_diagnostic_lines, I think we should probably rename it something like iter_diagnostic_position

i realise this can return a single diagnostic instead of a list of diagnostics, so i renamed it to `next_diagnostic` instead. What do you think?

> it might make sense to have an iter_diagnostic_lines function as well (instead of replacing the old one)

this function is internal and doesn't look like it's used anywhere else, so i'll prefer if we can avoid keeping the old one

> Maybe the new function should be iter_diagnostic_range, and instead of having a "search_forward" param have a start/end param that correspond to the buf position.

not sure if i understand this. The previous implementation supported wrapping by scanning through lines twice (eg. curr_line+1 to last_line, 1 to curr_line-1). I updated the implementation to use a offset+modulo to iterate through positions more easily, `search_forward` here is just a boolean switch to handle the various cases. For example, if we are searching forward and a diagnostic appears on the right of the same line, the nearest diagnostic to the right is returned

**Comment by pohzipohzi** [comment]
@mjlbach thanks for the quick review :) did some of the renaming as requested

**Comment by weilbith** [comment]
Thanks for "fixing" this! 🙏🏾

**Comment by pohzipohzi** [comment]
@mjlbach can you help to review again? I'll like to merge the changes before any conflicts arise :slightly_smiling_face: . Also this PR seems to have garnered some interest recently

**Comment by mjlbach** [comment]
Sorry for the delay, I'll try to review this again by next weekend.

File: runtime/lua/vim/lsp/diagnostic.lua
Comment (mjlbach) [inline runtime/lua/vim/lsp/diagnostic.lua:415]: Reminder to self to check client_get_diags is `ipair` save

File: runtime/lua/vim/lsp/diagnostic.lua
Comment (mjlbach) [inline runtime/lua/vim/lsp/diagnostic.lua:537]: These blocks seem like a lot of duplicated code, could turn into a small function that takes a comparison operator

File: runtime/lua/vim/lsp/diagnostic.lua
Comment (mfussenegger) [inline runtime/lua/vim/lsp/diagnostic.lua:1293]: This looks pretty much like a 1:1 copy of the `show_line_diagnostics` except that it misses improvements like done in https://github.com/neovim/neovim/commit/e43dbfdd05d09e28ee0468f1cd1808f859807606 and retrieves different diagnostics.

I think you could extract the common logic related to displaying the diagnostic into an internal `show_diagnostic` function that is parameterized with the diagnostics that should be displayed and then call that function both from `show_line_diagnostics` and from this new function.

File: runtime/lua/vim/lsp/diagnostic.lua
Comment (mfussenegger) [inline runtime/lua/vim/lsp/diagnostic.lua:391]: I wonder if we should instead extend `get` with a predicate function.

With `get`, `get_all`, `get_line_diagnostics` and now `get_position_diagnostics` it looks as if there's a trend to create a new function for each different set of diagnostics somebody wants to retrieve.

Comment (mjlbach) [reply runtime/lua/vim/lsp/diagnostic.lua:391]: +1, I initially suggested changes to get_position_diagnostics but I think it's way better to consolidate things and pass a predicate function depending on what slice of diagnostics you want

File: runtime/lua/vim/lsp/diagnostic.lua
Comment (pohzipohzi) [reply runtime/lua/vim/lsp/diagnostic.lua]: good suggestion, I've consolidated the shared parts of both functions into `show_diagnostics`, what do you think?

File: runtime/lua/vim/lsp/diagnostic.lua
Comment (pohzipohzi) [reply runtime/lua/vim/lsp/diagnostic.lua:391]: > we should instead extend get with a predicate function
i'm not sure what is being suggested here, can you give an example of how the consolidated function will look like?

Comment (mfussenegger) [reply runtime/lua/vim/lsp/diagnostic.lua:391]: For example, `M.get_all` could take an optional `predicate` as argument, that if present can be used to filter the arguments included.

You'd then call it within `show_position_diagnostics` with a predicate that filters the diagnostics by position:

```lua
local position_diagnostics = M.get_all(nil, function(diag)
  return (position[2] >= diag.range['start'].character)
    and (position[2] <= diag.range['end'].character or position[1] < diag.range['end'].line)
end)
```

File: runtime/lua/vim/lsp/diagnostic.lua
Comment (mfussenegger) [reply runtime/lua/vim/lsp/diagnostic.lua]: Looks good :+1

File: runtime/lua/vim/lsp/diagnostic.lua
Comment (pohzipohzi) [reply runtime/lua/vim/lsp/diagnostic.lua]: :+1: i tried to move the diagnostic retrieval logic for position diagnostics into `get`, what do you think?

File: runtime/lua/vim/lsp/diagnostic.lua
Comment (mfussenegger) [inline runtime/lua/vim/lsp/diagnostic.lua:369]: I'd leave out the `opts` parameter. The predicate can be used to filter based on the severity - there is no need to have a special case it.

I'd also remove the `comp` parameter - given that it is at the end, the call-site can do the `table.sort` call on the result - no need to inject the logic into `M.get`.

File: runtime/lua/vim/lsp/diagnostic.lua
Comment (mfussenegger) [inline runtime/lua/vim/lsp/diagnostic.lua:393]: It would also be necessary to apply the predicate to the contents of `diagnostic_cache` further below

File: runtime/lua/vim/lsp/diagnostic.lua
Comment (mfussenegger) [reply runtime/lua/vim/lsp/diagnostic.lua:369]: Another alternative might be to just keep `M.get()` as is, and do a `vim.tbl_filter(predicate, M.get(..))` at the call-site.
In some cases it may require an additional copy to create the filtered table, but not sure if that is a concern.

@mjlbach wdyt?

Comment (mjlbach) [reply runtime/lua/vim/lsp/diagnostic.lua:369]: I like your suggestion better @mfussenegger because it simplifies the interfacet to M.get but I don't know offhand if the perf characteristics would be worse given this is a large copy

Comment (pohzipohzi) [reply runtime/lua/vim/lsp/diagnostic.lua]: @mfussenegger I've removed opts and comp from the function signature but left `predicate` as a parameter :slightly_smiling_face: - lmk if this works. I can delegate the predicate logic to the call-site as well if you prefer

File: runtime/lua/vim/lsp/diagnostic.lua
Comment (mfussenegger) [inline runtime/lua/vim/lsp/diagnostic.lua:1261]:
```suggestion
  local match_position_predicate = function(diag)
    return position[1] == diag.range['start'].line and
    position[2] >= diag.range['start'].character and
    (position[2] <= diag.range['end'].character or position[1] < diag.range['end'].line)
  end
  local position_diagnostics = M.get(buf_nr, nil, match_position_predicate)
```

**Review by mfussenegger — COMMENTED** [review]
Looks like this comment https://github.com/neovim/neovim/pull/14795/files#r663566037 is still unaddressed? Otherwise I think this looks good.

Sorry for the late response. This somehow slipped through.

File: runtime/lua/vim/lsp/diagnostic.lua
Comment (pohzipohzi) [reply runtime/lua/vim/lsp/diagnostic.lua:494]: @mjlbach i did a small refactor based on your feedback :slightly_smiling_face: , what do you think?

File: runtime/lua/vim/lsp/diagnostic.lua
Comment (gpanders) [inline runtime/lua/vim/lsp/diagnostic.lua:1307]: This `--` should be a `---` or this line won't be rendered in the docs.

File: runtime/lua/vim/lsp/diagnostic.lua
Comment (gpanders) [inline runtime/lua/vim/lsp/diagnostic.lua:365]: In most other places in the docs we use `(number, optional)` to indicate an optional parameter (rather than `number|nil`. For the sake of consistency I'd suggest you stick with that.

Comment (pohzipohzi) [reply runtime/lua/vim/lsp/diagnostic.lua:365]: hey thanks for the review. The current state of docs seem a little all over the place now. When i tried to grep the `runtime/lua` directory i'm seeing 2 variants: `(optional, <type>)` and `(<type>, optional)`. This file however has a lot of `<type>|nil`, and I chose to use `<type>|nil` specifically because the line directly above already uses that

I'm thinking that this PR is already "consistent" in the sense that it follows the conventions in its surrounding code. Since there are so many inconsistencies we can instead have a separate PR which consolidates the different type definition formats in one go. What do you think? Just my two cents, we can still fix the above types if you insist

Comment (gpanders) [reply runtime/lua/vim/lsp/diagnostic.lua:365]: Staying consistent with the nearest code has highest priority I think, so this is fine. Thanks for the explanation.

**Comment by mjlbach** [comment]
With the changes this LGTM, happy to merge once @mfussenegger gives approval

**Review by mfussenegger — APPROVED** [review]
looks all good to me now :+1:

**Pull request merged**
